### PR TITLE
Add Andersen Alias Analysis pass

### PR DESCRIPTION
**Pull request by haved**
Builds on top of #298 and #299

Adds an Andersen pass, creating `PointerObject`s and constraints for a given rvsdg-llvm-module, solving the constraints, and producing a `PointsToGraph`.

All tests fro `TestSteensgaard.cpp` are ported over to `TestAndersen.cpp`, except for the Statistics test. The solutions are of course a bit different, mostly removing edges due to more precise analysis, but also adding some, due to Steensgaard not registering imports and lambdas as escaped / external.

### CODE REVIEW

File: jlm/llvm/opt/alias-analyses/Andersen.hpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.hpp:22]: Could you provide a reference to the paper such that people know which one it is that you implemented? Similarly to [Steensgaard](https://github.com/phate/jlm/blob/f2e1a788b57d40b0438223e195d9afd776b1dbd8/jlm/llvm/opt/alias-analyses/Steensgaard.hpp#L15C1-L23C47).

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:23]: Could you use [is()](https://github.com/phate/jlm/blob/f2e1a788b57d40b0438223e195d9afd776b1dbd8/jlm/rvsdg/node.hpp#L903C1-L911C2) here? It would be more readable.

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:28]:
```
else if (loadNode = dynamic_cast<const LoadNode*>(&node))
  AnalyzeLoad(*loadNode);
```

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:102]: "be conservative and maRK the address"?

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:233]: It should be ConstantArray, not ConstantStruct, no? Considering that this not trigger anything, I suspect that there is no unit test that covers this function either?

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:313]: I believe you should use `dynamic_cast<>()` on the node and then throw out the AssertedCast.

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:338]: The comment is not very helpful. The code line below says what it does.

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:357]: Handle

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:369]: same here.

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:416]: same here

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:454]: same here

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:480]: `for (auto thetaOutput : theta)` maybe?

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:498]: same here.

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:658]: This check does not do anything, no? Should it have been `JLM_ASSERT(targetIdx < memoryNodes.size())`?

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:675]: The same as above. This is undefined behavior if idx is > memoryNodes.size()? At least according to the notes on this page: https://en.cppreference.com/w/cpp/container/vector/operator_at

Or do I miss something?

File: tests/jlm/llvm/opt/alias-analyses/TestAndersen.cpp
Comment (phate) [inline tests/jlm/llvm/opt/alias-analyses/TestAndersen.cpp:230]: Please use `//` instead of `/* ... */`

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (haved) [reply jlm/llvm/opt/alias-analyses/Andersen.cpp:658]: It does, memoryNodes is nullptr for all indexes that correspond to PointerObjects of Register kind

File: jlm/llvm/opt/alias-analyses/Andersen.cpp
Comment (haved) [reply jlm/llvm/opt/alias-analyses/Andersen.cpp:675]: memoryNodes is made to be as long as `NumPointerObjects()`, to make finding PtG nodes corresponding to a PointerObject fast.

File: tests/jlm/util/TestBiMap.cpp
Comment (phate) [inline tests/jlm/util/TestBiMap.cpp:1]: This file can disappear from the PR, no?

File: jlm/util/BiMap.hpp
Comment (phate) [inline jlm/util/BiMap.hpp:1]: This file can disappear from the PR, no?

**Review by phate — CHANGES_REQUESTED** [review]
The two files still need to disappear, otherwise LGTM.

**Comment by haved** [comment]
Oh oops, absolutely, git merge happily kept both versions. I forgot to merge with the non-squashed version of the last PR before merging with master.

**Review by phate — APPROVED** [review]
LGTM

**Pull request merged**
